lib/spandex_ecto/ecto_logger.ex

```diff
@@ -18,75 +24,111 @@ defmodule SpandexEcto.EctoLogger do
+      |> String.slice(0, truncate)
+
+    num_rows = num_rows(log_entry)
+
+    queue_time = get_time(log_entry, :queue_time)
+    query_time = get_time(log_entry, :query_time)
+    decoding_time = get_time(log_entry, :decode_time)
+
+    start = now - (queue_time + query_time + decoding_time)
+
+    tracer.update_span(
+      start: start,
+      completion_time: now,
+      service: service,
+      resource: query,
+      type: :db,
+      sql_query: [
+        query: query,
+        rows: inspect(num_rows),
+        db: database
+      ],
+      tags: tags(log_entry)
+    )
+
+    report_error(tracer, log_entry)
+
+    if queue_time != 0 do
+      tracer.start_span("queue")
+      tracer.update_span(service: service, start: start, completion_time: start + queue_time)
+      tracer.finish_span()
+    end
 
-      start = now - (queue_time + query_time + decoding_time)
+    if query_time != 0 do
+      tracer.start_span("run_query")
 
-      tracer.start_span(
-        "query",
-        start: start,
-        completion_time: now,
+      tracer.update_span(
         service: service,
-        resource: query,
-        type: :db,
-        sql_query: [
-          query: query,
-          rows: inspect(num_rows),
-          db: database
-        ],
-        tags: tags(log_entry)
+        start: start + queue_time,
+        completion_time: start + queue_time + query_time
       )
 
-      Logger.metadata(trace_id: tracer.current_trace_id(), span_id: tracer.current_span_id())
+      tracer.finish_span()
+    end
 
-      report_error(tracer, log_entry)
+    if decoding_time != 0 do
+      tracer.start_span("decode")
 
-      if queue_time != 0 do
-        tracer.start_span("queue")
-        tracer.update_span(service: service, start: start, completion_time: start + queue_time)
-        tracer.finish_span()
-      end
+      tracer.update_span(
+        service: service,
+        start: start + queue_time + query_time,
+        completion_time: now
+      )
+
+      tracer.finish_span()
+    end
 
-      if query_time != 0 do
-        tracer.start_span("run_query")
+    finish_ecto_trace(caller_pid, tracer)
 
-        tracer.update_span(
-          service: service,
-          start: start + queue_time,
-          completion_time: start + queue_time + query_time
-        )
+    log_entry
+  end
 
-        tracer.finish_span()
+  defp finish_ecto_trace(caller_pid, tracer) do
+    if caller_pid != self() do
+      tracer.finish_trace()
+    else
+      tracer.finish_span()
+    end
+  end
+
+  defp setup(caller_pid, tracer) when is_pid(caller_pid) do
+    if caller_pid == self() do
+      if tracer.current_trace_id() do
+        tracer.start_span("query")
       end
+    else
+      dictionary_map = Process.info(caller_pid)[:dictionary] |> Enum.into(%{})
 
-      if decoding_time != 0 do
-        tracer.start_span("decode")
+      case dictionary_map[{:spandex_trace, tracer}] do
+        nil ->
+          tracer.start_span("query")
```

Comment:
If you are starting span without ongoing trace this will return an error, no span will be started. I think this clause should be removed at all